core/controllers/editor.py

```diff
@@ -613,59 +613,83 @@ class ImageUploadHandler(EditorHandler):
+
+            # Verify that the data is recognized as an image.
+            file_format = imghdr.what(None, h=raw)
+            if file_format not in feconf.ACCEPTED_IMAGE_FORMATS_AND_EXTENSIONS:
+                raise self.InvalidInputException('Image not recognized')
+
+            # Verify that the file type matches the supplied extension.
+            if not filename:
+                raise self.InvalidInputException('No filename supplied')
+            if filename.rfind('.') == 0:
+                raise self.InvalidInputException('Invalid filename')
+            if '/' in filename or '..' in filename:
+                raise self.InvalidInputException(
+                    'Filenames should not include slashes (/) or consecutive '
+                    'dot characters.')
+            if '.' not in filename:
+                raise self.InvalidInputException(
+                    'Image filename with no extension: it should have '
+                    'one of the following extensions: %s.' % allowed_formats)
 
-        dot_index = filename.rfind('.')
-        extension = filename[dot_index + 1:].lower()
-        if (extension not in
-                feconf.ACCEPTED_IMAGE_FORMATS_AND_EXTENSIONS[file_format]):
-            raise self.InvalidInputException(
-                'Expected a filename ending in .%s, received %s' %
-                (file_format, filename))
+            dot_index = filename.rfind('.')
+            extension = filename[dot_index + 1:].lower()
+            if (extension not in
+                    feconf.ACCEPTED_IMAGE_FORMATS_AND_EXTENSIONS[file_format]):
+                raise self.InvalidInputException(
+                    'Expected a filename ending in .%s, received %s' %
+                    (file_format, filename))
 
-        file_system_class = fs_services.get_exploration_file_system_class()
-        fs = fs_domain.AbstractFileSystem(file_system_class(
-            fs_domain.ENTITY_TYPE_EXPLORATION, exploration_id))
-        filepath = '%s/%s' % (self._FILENAME_PREFIX, filename)
+            file_system_class = fs_services.get_entity_file_system_class()
+            fs = fs_domain.AbstractFileSystem(file_system_class(
+                entity_type, entity_id))
+            filepath = '%s/%s' % (self._FILENAME_PREFIX, filename)
 
-        if fs.isfile(filepath):
-            raise self.InvalidInputException(
-                'A file with the name %s already exists. Please choose a '
-                'different name.' % filename)
+            if fs.isfile(filepath):
+                raise self.InvalidInputException(
+                    'A file with the name %s already exists. Please choose a '
+                    'different name.' % filename)
 
-        exp_services.save_original_and_compressed_versions_of_image(
-            self.user_id, filename, exploration_id, raw)
+            fs_services.save_original_and_compressed_versions_of_image(
+                self.user_id, filename, entity_id, entity_type, raw)
```

Comment:
Always put entity_type before entity_id, it makes more logical sense that way.

Reply:
Done